string type
float32 distance